File: Project_code/OP_PDFS.cpp
#include <iostream>
#include <vector>
#include <list>
#include <algorithm>

using namespace std;

// Structure to represent an edge in the tree
struct Edge
{
    int to;
    double weight;
};

// Structure to represent a tree
class Tree
{
private:
    int n;                    // Number of vertices
    int root;                 // Root of the tree
    vector<vector<Edge>> adj; // Adjacency list
    vector<bool> visited;     // To track visited nodes during DFS

    // Helper function for DFS traversal
    void dfs(int v, vector<int> &rdfs)
    {
        visited[v] = true;
        rdfs.push_back(v);

        for (const Edge &e : adj[v])
        {
            if (!visited[e.to])
            {
                dfs(e.to, rdfs);
                rdfs.push_back(v); // Return to parent after exploring subtree
            }
        }
    }

public:
    Tree(int n, int root) : n(n), root(root)
    {
        adj.resize(n);
        visited.resize(n, false);
    }

    // Add an undirected edge to the tree
    void addEdge(int u, int v, double weight)
    {
        adj[u].push_back({v, weight});
        adj[v].push_back({u, weight});
    }

    // Generate the RDFS traversal
    vector<int> generateRDFS()
    {
        vector<int> rdfs;
        fill(visited.begin(), visited.end(), false);
        dfs(root, rdfs);
        return rdfs;
    }

    // Calculate distance between two vertices
    double distance(int u, int v)
    {
        if (u == v)
            return 0;

        vector<bool> vis(n, false);
        vector<double> dist(n, -1);
        list<int> queue;

        vis[u] = true;
        dist[u] = 0;
        queue.push_back(u);

        while (!queue.empty())
        {
            int s = queue.front();
            queue.pop_front();

            for (const Edge &e : adj[s])
            {
                if (!vis[e.to])
                {
                    vis[e.to] = true;
                    dist[e.to] = dist[s] + e.weight;

                    if (e.to == v)
                        return dist[e.to];

                    queue.push_back(e.to);
                }
            }
        }

        return -1; // Should not reach here in a connected tree
    }

    // Calculate the length of a path
    double pathLength(const vector<int> &path)
    {
        double length = 0;
        for (size_t i = 1; i < path.size(); i++)
        {
            // Find the edge weight between path[i-1] and path[i]
            for (const Edge &e : adj[path[i - 1]])
            {
                if (e.to == path[i])
                {
                    length += e.weight;
                    break;
                }
            }
        }
        return length;
    }

    // Find the path from u to v
    vector<int> findPath(int u, int v)
    {
        if (u == v)
            return {u};

        vector<int> parent(n, -1);
        vector<bool> vis(n, false);
        list<int> queue;

        vis[u] = true;
        queue.push_back(u);

        while (!queue.empty())
        {
            int s = queue.front();
            queue.pop_front();

            for (const Edge &e : adj[s])
            {
                if (!vis[e.to])
                {
                    vis[e.to] = true;
                    parent[e.to] = s;

                    if (e.to == v)
                    {
                        // Construct the path
                        vector<int> path;
                        int current = v;
                        while (current != -1)
                        {
                            path.push_back(current);
                            current = parent[current];
                        }
                        reverse(path.begin(), path.end());
                        return path;
                    }

                    queue.push_back(e.to);
                }
            }
        }

        return {}; // Should not reach here in a connected tree
    }

    // Calculate the accumulated weight for a path
    double getAccumulatedWeight(const vector<int> &path)
    {
        double totalWeight = 0;
        for (size_t i = 1; i < path.size(); i++)
        {
            for (const Edge &e : adj[path[i - 1]])
            {
                if (e.to == path[i])
                {
                    totalWeight += e.weight;
                    break;
                }
            }
        }
        return totalWeight;
    }
};

// Structure to represent a route
struct Route
{
    vector<int> vertices;
    double length;
};

// Print a route
void printRoute(const Route &route)
{
    cout << "Route: ";
    for (int vertex : route.vertices)
    {
        cout << vertex << " ";
    }
    cout << "Length: " << route.length << endl;
}

void inputTree(Tree &tree, int n)
{
    for (int i = 0; i < n - 1; i++)
    {
        int u, v;
        double weight;
        cout << "Enter edge (u, v) and weight: ";
        cin >> u >> v >> weight;
        tree.addEdge(u, v, weight);
    }
}

// Calculate PDFS strategy
vector<Route> calculatePDFS(Tree &tree, double B)
{
    vector<Route> routes;
    vector<int> rdfs = tree.generateRDFS();
    int root = rdfs[0];
    int n = rdfs.size();
    vector<bool> visited(n, false);
    visited[root] = true;

    // Tạo danh sách các đỉnh và trọng số tương ứng từ root
    vector<pair<int, double>> vertices;
    for (int i = 0; i < n; i++)
    {
        if (i != root)
        {
            double dist = tree.distance(root, i);
            if (dist > 0) // Chỉ xét các đỉnh có kết nối
            {
                vertices.push_back({i, dist});
            }
        }
    }

    // Sắp xếp các đỉnh theo trọng số tăng dần
    sort(vertices.begin(), vertices.end(),
         [](const pair<int, double> &a, const pair<int, double> &b)
         {
             return a.second < b.second;
         });

    Route currentRoute;
    currentRoute.vertices = {root};
    double currentLength = 0;

    // Nhóm các đỉnh theo trọng số để tạo các tuyến đường
    while (!vertices.empty())
    {
        vector<int> currentGroup;
        double targetWeight = vertices[0].second; // Trọng số của đỉnh đầu tiên chưa thăm

        // Tìm các đỉnh có trọng số gần bằng nhau để gom vào một nhóm
        for (const auto &v : vertices)
        {
            if (!visited[v.first] && v.second <= targetWeight * 1.5) // Độ chênh lệch cho phép 50%
            {
                vector<int> tempPath = currentRoute.vertices;

                // Thử thêm đỉnh vào tuyến đường hiện tại
                for (int vertex : currentGroup)
                {
                    vector<int> pathToVertex = tree.findPath(tempPath.back(), vertex);
                    for (size_t i = 1; i < pathToVertex.size(); i++)
                    {
                        tempPath.push_back(pathToVertex[i]);
                    }
                }

                vector<int> pathToNew = tree.findPath(tempPath.back(), v.first);
                vector<int> returnToRoot = tree.findPath(v.first, root);

                vector<int> testPath = tempPath;
                for (size_t i = 1; i < pathToNew.size(); i++)
                {
                    testPath.push_back(pathToNew[i]);
                }
                for (size_t i = 1; i < returnToRoot.size(); i++)
                {
                    testPath.push_back(returnToRoot[i]);
                }

                double testLength = tree.pathLength(testPath);
                if (testLength <= B)
                {
                    currentGroup.push_back(v.first);
                }
            }
        }

        if (!currentGroup.empty())
        {
            Route route;
            route.vertices = {root};

            // Xây dựng tuyến đường cho nhóm hiện tại
            for (int vertex : currentGroup)
            {
                vector<int> pathToVertex = tree.findPath(route.vertices.back(), vertex);
                for (size_t i = 1; i < pathToVertex.size(); i++)
                {
                    route.vertices.push_back(pathToVertex[i]);
                }
            }

            // Thêm đường về root
            if (route.vertices.back() != root)
            {
                vector<int> returnPath = tree.findPath(route.vertices.back(), root);
                for (size_t i = 1; i < returnPath.size(); i++)
                {
                    route.vertices.push_back(returnPath[i]);
                }
            }

            route.length = tree.pathLength(route.vertices);
            if (route.length <= B)
            {
                routes.push_back(route);
                // Đánh dấu các đỉnh đã thăm
                for (int vertex : currentGroup)
                {
                    visited[vertex] = true;
                }
            }
        }

        // Loại bỏ các đỉnh đã thăm khỏi danh sách
        vertices.erase(
            remove_if(vertices.begin(), vertices.end(),
                      [&visited](const pair<int, double> &v)
                      { return visited[v.first]; }),
            vertices.end());
    }

    return routes;
}

int main()
{
    int n; // Number of vertices
    int root;
    double B; // Energy budget
    cout << "Enter the number of vertices in the tree: ";
    cin >> n;
    cout << "Enter the energy budget B: ";
    cin >> B;
    cout << "Enter the root of the tree: ";
    cin >> root;
    if (n <= 0 || B <= 0)
    {
        cerr << "Invalid input. Number of vertices and energy budget must be positive." << endl;
        return 1;
    }
    // Create a tree similar to the one in Fig.1 of the problem statement
    Tree tree(n, root);
    cout << "Input the edges of the tree:" << endl;
    inputTree(tree, n);

    vector<Route> pdfs = calculatePDFS(tree, B);

    // Print the routes
    cout << "PDFS Strategy with B = " << B << ":" << endl;
    for (size_t i = 0; i < pdfs.size(); i++)
    {
        cout << "R" << i + 1 << ": ";
        printRoute(pdfs[i]);
    }

    cout << "Total number of routes: " << pdfs.size() << endl;
}

File: Project_code/OP_PDFS.py
from dataclasses import dataclass
from typing import List, Tuple, Dict
from collections import deque


@dataclass
class Edge:
    to: int
    weight: float


class Tree:
    def __init__(self, n: int, root: int):
        self.n = n
        self.root = root
        self.adj = [[] for _ in range(n)]
        self.visited = [False] * n

    def add_edge(self, u: int, v: int, weight: float):
        self.adj[u].append(Edge(v, weight))
        self.adj[v].append(Edge(u, weight))

    def _dfs(self, v: int, rdfs: List[int]):
        self.visited[v] = True
        rdfs.append(v)

        for e in self.adj[v]:
            if not self.visited[e.to]:
                self._dfs(e.to, rdfs)
                rdfs.append(v)

    def generate_rdfs(self) -> List[int]:
        self.visited = [False] * self.n
        rdfs = []
        self._dfs(self.root, rdfs)
        return rdfs

    def distance(self, u: int, v: int) -> float:
        if u == v:
            return 0.0

        vis = [False] * self.n
        dist = [-1.0] * self.n
        queue = deque([u])
        vis[u] = True
        dist[u] = 0.0

        while queue:
            s = queue.popleft()
            for e in self.adj[s]:
                if not vis[e.to]:
                    vis[e.to] = True
                    dist[e.to] = dist[s] + e.weight
                    if e.to == v:
                        return dist[e.to]
                    queue.append(e.to)
        return -1.0

    def path_length(self, path: List[int]) -> float:
        length = 0.0
        for i in range(1, len(path)):
            for e in self.adj[path[i - 1]]:
                if e.to == path[i]:
                    length += e.weight
                    break
        return length

    def find_path(self, u: int, v: int) -> List[int]:
        if u == v:
            return [u]

        parent = [-1] * self.n
        vis = [False] * self.n
        queue = deque([u])
        vis[u] = True

        while queue:
            s = queue.popleft()
            for e in self.adj[s]:
                if not vis[e.to]:
                    vis[e.to] = True
                    parent[e.to] = s
                    if e.to == v:
                        path = []
                        current = v
                        while current != -1:
                            path.append(current)
                            current = parent[current]
                        return path[::-1]
                    queue.append(e.to)
        return []

    def get_accumulated_weight(self, path: List[int]) -> float:
        total_weight = 0.0
        for i in range(1, len(path)):
            for e in self.adj[path[i - 1]]:
                if e.to == path[i]:
                    total_weight += e.weight
                    break
        return total_weight


@dataclass
class Route:
    vertices: List[int]
    length: float


def calculate_pdfs(tree: Tree, B: float) -> List[Route]:
    routes = []
    rdfs = tree.generate_rdfs()
    root = rdfs[0]
    n = len(rdfs)
    visited = [False] * n
    visited[root] = True

    # Create list of vertices and their weights from root
    vertices = []
    for i in range(n):
        if i != root:
            dist = tree.distance(root, i)
            if dist > 0:  # Only consider connected vertices
                vertices.append((i, dist))

    # Sort vertices by weight
    vertices.sort(key=lambda x: x[1])

    current_route = Route(vertices=[root], length=0.0)

    while vertices:
        current_group = []
        target_weight = vertices[0][1]  # Weight of first unvisited vertex

        # Find vertices with similar weights to group together
        for v, weight in vertices:
            if not visited[v] and weight <= target_weight * 1.5:  # Allow 50% difference
                temp_path = current_route.vertices.copy()

                # Try adding vertex to current route
                for vertex in current_group:
                    path_to_vertex = tree.find_path(temp_path[-1], vertex)
                    temp_path.extend(path_to_vertex[1:])

                path_to_new = tree.find_path(temp_path[-1], v)
                return_to_root = tree.find_path(v, root)

                test_path = temp_path.copy()
                test_path.extend(path_to_new[1:])
                test_path.extend(return_to_root[1:])

                test_length = tree.path_length(test_path)
                if test_length <= B:
                    current_group.append(v)

        if current_group:
            route = Route(vertices=[root], length=0.0)

            # Build route for current group
            for vertex in current_group:
                path_to_vertex = tree.find_path(route.vertices[-1], vertex)
                route.vertices.extend(path_to_vertex[1:])

            # Add return path to root
            if route.vertices[-1] != root:
                return_path = tree.find_path(route.vertices[-1], root)
                route.vertices.extend(return_path[1:])

            route.length = tree.path_length(route.vertices)
            if route.length <= B:
                routes.append(route)
                # Mark visited vertices
                for vertex in current_group:
                    visited[vertex] = True

        # Remove visited vertices from list
        vertices = [(v, w) for v, w in vertices if not visited[v]]

    return routes


def print_route(route: Route, index: int = None):
    route_str = f"R{index}: " if index is not None else "Route: "
    vertices_str = " ".join(str(v) for v in route.vertices)
    print(f"{route_str}{vertices_str} Length: {route.length}")


def input_tree(tree: Tree, n: int):
    for i in range(n - 1):
        print("Enter edge (u, v) and weight: ", end="")
        u, v, weight = map(float, input().split())
        tree.add_edge(int(u), int(v), weight)


def main():
    print("Enter the number of vertices in the tree: ", end="")
    n = int(input())
    print("Enter the energy budget B: ", end="")
    B = float(input())
    print("Enter the root of the tree: ", end="")
    root = int(input())

    if n <= 0 or B <= 0:
        print("Invalid input. Number of vertices and energy budget must be positive.")
        return

    tree = Tree(n, root)
    print("Input the edges of the tree:")
    input_tree(tree, n)

    pdfs = calculate_pdfs(tree, B)

    print(f"\nPDFS Strategy with B = {B}:")
    for i, route in enumerate(pdfs, 1):
        print_route(route, i)

    print(f"Total number of routes: {len(pdfs)}")


if __name__ == "__main__":
    main()
